internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,297 @@
+	return c.tp.Tracer("cloudquery.io/internal/telemetry")
+}
+
+func (c *Client) Shutdown(ctx context.Context) {
+	if err := c.HasError(); err != nil {
+		c.logger.Debug("telemetry error", "error", err)
+	}
+
+	if c.disabled {
+		return
+	}
+
+	if sd, ok := c.tp.(shutdownable); ok {
+		if err := sd.Shutdown(ctx); err != nil {
+			c.logger.Debug("shutdown failed", "error", err)
+		}
+	}
+
+	if c.closer != nil {
+		if err := c.closer.Close(); err != nil {
+			c.logger.Debug("close failed", "error", err)
+		}
+	}
+}
+
+// RecordError should be called on a span to mark it as errored. Error values are not included unless debug mode is on.
+func (c *Client) RecordError(span otrace.Span, err error, opts ...otrace.EventOption) {
+	if err == nil {
+		return
+	}
+
+	if c.debug {
+		span.RecordError(err, opts...)
+		span.SetStatus(codes.Error, err.Error())
+		return
+	}
+
+	span.RecordError(fmt.Errorf("error"))
+	span.SetStatus(codes.Error, "error")
+}
+
+func (c *Client) HasError() error {
+	return c.err
+}
+
+func (c *Client) setError(err error) {
+	if err != nil {
+		c.logger.Debug("telemetry error occurred", "error", err)
+	}
+	if c.err == nil && err != nil {
+		c.err = err
+	}
+}
+
+func (c *Client) defaultResource(ctx context.Context) (*resource.Resource, error) {
+	cookieContents, err := c.cookie()
+	if err != nil {
+		c.logger.Debug("cookie failed", "error", err)
+	}
+
+	attr := []attribute.KeyValue{
+		semconv.ServiceNameKey.String("cloudquery"),
+		semconv.ServiceVersionKey.String(c.version),
+		attribute.String("commit", c.commit),
+		attribute.String("build_date", c.buildDate),
+		attribute.Bool("ci", isCI()),
+	}
+	if cookieContents != "" {
+		attr = append(attr, semconv.ServiceInstanceIDKey.String(cookieContents))
+	}
+	if hn, err := os.Hostname(); err == nil && hn != "" {
+		attr = append(attr, semconv.HostNameKey.String(hashAttribute(hn)))
+	}
+	attr = append(attr, osInfo()...)
+
+	return resource.New(ctx,
+		resource.WithTelemetrySDK(),
+		resource.WithProcessRuntimeName(),
+		resource.WithProcessRuntimeVersion(),
+		resource.WithProcessRuntimeDescription(),
+		resource.WithAttributes(attr...),
+	)
+}
+
+func (c *Client) cookie() (string, error) {
+	fn := filepath.Join(".", ".cq", "telemetry-cookie")
+
+	exists := true
+	fi, err := c.fs.Stat(fn)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return "", err
+		}
+		exists = false
+	}
+	if exists && fi.IsDir() {
+		return "", fmt.Errorf("telemetry-cookie is a directory")
+	}
+
+	if exists {
+		b, err := c.fs.ReadFile(fn)
+		if err != nil {
+			return "", err
+		}
+		return string(b), nil
+	}
+
+	id := uuid.NewV4().String()
+	if err := c.fs.WriteFile(fn, []byte(id), 0644); err != nil {
+		return "", err
+	}
+	return id, nil
+}
+
+func (c *Client) defaultExporter(ctx context.Context) (trace.SpanExporter, error) {
+	return otlptracegrpc.New(
+		ctx,
+		otlptracegrpc.WithInsecure(), // TODO change
+		otlptracegrpc.WithEndpoint("localhost:4317"), // TODO change. env var?
```

Comment:
I would say this should be hardcoded, maybe make it configurable in the future so we can support proxies